### ISSUE TITLE
Run a profiler 

### ISSUE THREAD

**Issue by mishmosh** [filecoin-project/go-filecoin#2149]
### Description
We believe there are low-hanging performance and memory fruit all over go-filecoin. It would be great if someone with experience in profiling could take a preliminary look, and write up their results.

### Acceptance criteria
- Write-up of results & assessment

**Comment by mishmosh** [comment]
@phritz @whyrusleeping anything else to add?

**Comment by ZenGround0** [comment]
FYI the daemon already runs a [pprof server](https://github.com/filecoin-project/go-filecoin/blob/320997b4f007cc655bcaf11be82b26f34a6b41ed/commands/daemon.go#L163) on the commands api port, so curling `localhost:3453/debug/pprof/` is a good place to start.

curling that is the place to start

**Comment by whyrusleeping** [comment]
Guide i've written on how to start with this: https://github.com/ipfs/go-ipfs/blob/master/docs/debug-guide.md

I was looking through these profiles recently and picked up a bunch of low hanging fruit. There is definitely more, this should be fun for whoever wants to pick it up.

**Comment by mishmosh** [comment]
@mslipper loves memory profiling and is going to get started on this

**Comment by mslipper** [comment]
Ok, so I started out by profiling memory usage during and after a full sync. Here are the high-level things that I found:

1. The single largest user of memory is the Rust groth param cache. Memory usage jumped from 460MB to 2.4GB as soon as those were loaded. Note that this was inferred from the daemon's total memory usage since `pprof` includes only the Go heap in its output.
2. In second place is the ~330MB block of memory Badger allocates for its internal skip list data structure. >70% of Go's heap is devoted to this single structure. This value is configurable via the `MaxTableSize` and `MaxNodeSize` options. I'd recommend we look at tuning these, since this value is 10% of the entire on-disk database size. (Separately, I can't find where Badger's garbage collector is called - the `CollectGarbage` method doesn't appear to be used anywhere in the `go-filecoin` codebase.)

The above are the results of the `inuse_space` profiler. When I ran the `alloc_space` profiler, things were a bit different. The only obvious thing I saw here was an over-allocation of `big.Int` objects. The culprit seems to be in `go-leb128`, which allocates several `big.Int` objects as part of an inner loop in `FromBigInt`. I'll have a go at optimizing this loop a bit to reduce interim `big.Int` allocations. These account for ~7% of all allocations.

**Comment by phritz** [comment]
thanks @mslipper ! the groth thing is a known issue, unavoidable at this point. i dont know much about badger but if there is something we should be doing but are not, please feel free to propose a fix.

re leb128 if you have an optimization handy, great. however it is possible that we're going to move away from leb128 in the future so i would not spend too much more time on it until we're certain. 

i would love to see what cpu profiling turns up, is that in scope for you?

thanks for doing this!

**Comment by mslipper** [comment]
CPU profiling is definitely in scope! I performed one today. Here's what I discovered:

- Initially syncing `go-filecoin` uses about 30% of a core, but I didn't see anything in the profile that was noteworthy since the sync process is so network-bound. Most of the CPU cycles were spent serializing/deserializing CBOR objects, and reading/writing network data.
- Since there wasn't anything notable there, I ran a profile of the production `go-filecoin` daemon feeding stats to the dashboard. This node receives lots of API requests while remaining in sync - `chain.Ls()` and `miner.Power()` requests, respectively - and generally pegs 1 core completely. Here, I discovered that approximately 10% of all CPU time was spent in `leb128.FromBigInt`. The majority of the rest was spent performing `reflect` calls in `go-filecoin`'s various CBOR libraries.

Given the above, I'm going to try my hand at optimizing `leb128`. I don't have a _specific_ optimization in mind, but I believe it'd be worth it for me to come up with one since the impact seems high.

Here's a link to the CPU profile: [profile.speedscope.json.zip](https://github.com/filecoin-project/go-filecoin/files/2973560/profile.speedscope.json.zip).

(Side note: There's a pretty great profile visualizer tool I found called [speedscope.app](https://speedscope.app) that lets you 'left align' function calls by time spent. It made finding the `leb128` issue pretty quick.)
